library/certificate_request.py

```diff
@@ -0,0 +1,119 @@
+# -*- coding: utf-8 -*-
+
+ANSIBLE_METADATA = {
+    "metadata_version": "1.1",
+    "status": ["preview"],
+    "supported_by": "community",
+}
+
+DOCUMENTATION = """
+---
+module: certificate_request
+short_description: Manage SSL/TLS certificates.
+description:
+  - The C(certificate_request) module takes a name, desired
+    certificate request attributes and certificate properties.
+  - The request is generated and sent to the CA to sign.
+
+options:
+  name:
+    description:
+      - Name of the certificate. A full path can be used
+        to choose the directory where files will be stored.
+    required: true
+  dns:
+    description:
+      - Domain (or list of domains) to be included in the
+        certificate. Also can provide the default value for
+        common_name.
+    required: false
+  ca:
+    description:
+      - CA that will issue the certificate. The available options
+        will vary depending on each provider.
+    required: true
+  provider:
+    description:
+      - The underlying method used to request and manage the
+        certificate.
+    required: false
+    default: certmonger
+author:
+  - Sergio Oliveira Campos (@seocam)
+"""
+
+EXAMPLES = """
+# Certificate for single domain
+- name: Ensure certificate exists for www.example.com
+  certificate_request:
+    name: single-example
+    dns: www.example.com
+    ca: self-sign
+
+# Certificate for multiple domains
+- name: Ensure certificate exists for multiple domains
+  certificate_request:
+    name: many-example
+    dns:
+      - www.example.com
+      - example.com
+    ca: self-sign
+"""
+
+RETURN = ""
+
+from ansible.module_utils.basic import AnsibleModule
+from ansible.module_utils.certificate.providers import certmonger
+
+
+class CertificateRequestModule(AnsibleModule):
+    """Certificate Request Module.
+
+    This module is responsible for converging SSL/TLS certificates
+    using providers.
+    """
+
+    # Maps a provider name to a provider class
+    _providers_map = {
+        "certmonger": certmonger.CertificateRequestCertmongerProvider,
+    }
+
+    def __init__(self, *args, **kwargs):
+        argument_spec = kwargs.get("argument_spec", {})
+        argument_spec.update(self._get_argument_spec())
+
+        super(CertificateRequestModule, self).__init__(argument_spec, *args, **kwargs)
+
+    @staticmethod
+    def _get_argument_spec():
+        """Return a dict with the module arguments."""
+        return dict(
+            name=dict(type="str", required=True),
+            dns=dict(type="list"),
+            ca=dict(type="str"),
+            provider=dict(type="str", default="certmonger"),
+        )
+
+    @property
+    def provider(self):
```

Comment:
This isn't really a property, is it?  AFAICT, it will instantiate a new object of class `CertificateRequestXXXProvider` (where XXX is the provider type) every time it is called - that is, it doesn't get/set an object member called `_provider` or some such.

Reply:
That's true. I didn't get very much concerned about that because it should be called just once but that could change in the future. I'll change it.

Reply:
99b55cd97725b3448d8cbc9609a900ed94e45486